cro-clib/src/transaction_build.rs

```diff
@@ -0,0 +1,294 @@
+    let mut tx: Tx = Tx::default();
+    tx.inputs = user_tx.txin_pointer.clone();
+    tx.outputs = user_tx.txout.clone();
+    let mut access_policies = BTreeSet::new();
+    for viewkey_user in &user_tx.viewkey {
+        access_policies.insert(TxAccessPolicy {
+            view_key: viewkey_user.clone(),
+            access: TxAccess::AllData,
+        });
+    }
+    tx.attributes = TxAttributes::new_with_access(network, access_policies.into_iter().collect());
+    user_tx.tx = Some(tx);
+    Ok(())
+}
+
+/// prepare tx for signing
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_prepare_for_signing(tx_ptr: CroTxPtr, network: u8) -> CroResult {
+    let mut user_tx: &mut CroTx = tx_ptr.as_mut().expect("get tx");
+    match do_cro_tx_prepare_for_signing(&mut user_tx, network) {
+        Ok(_) => CroResult::success(),
+        Err(_) => CroResult::fail(),
+    }
+}
+
+/// extract bytes from singed tx
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_complete_signing(
+    tx_ptr: CroTxPtr,
+    output: *mut u8,
+    output_length: *mut u32,
+) -> CroResult {
+    let user_tx: &mut CroTx = tx_ptr.as_mut().expect("get tx");
+    let encoded: Vec<u8> = user_tx.tx.as_ref().expect("get tx core").encode();
+    ptr::copy_nonoverlapping(encoded.as_ptr(), output, encoded.len());
+    (*output_length) = encoded.len() as u32;
+    CroResult::success()
+}
+
+fn do_cro_tx_sign_txin(
+    address: &CroAddress,
+    user_tx: &mut CroTx,
+    which_tx_in_user: u16,
+) -> Result<()> {
+    let which_tx_in: usize = which_tx_in_user as usize;
+    assert!(user_tx.tx.is_some());
+    assert!(which_tx_in < user_tx.txin.len());
+    assert!(which_tx_in < user_tx.txin_witness.len());
+    let tx: &mut Tx = user_tx.tx.as_mut().chain(|| {
+        (
+            ErrorKind::DeserializationError,
+            "Unable to decode hex of txid",
+        )
+    })?;
+    let txid: TxId = tx.id();
+    let witness: TxInWitness = schnorr_sign(&txid, &address.publickey, &address.privatekey)?;
+    user_tx.txin_witness[which_tx_in] = Some(witness);
+    assert!(user_tx.txin.len() == user_tx.txin_witness.len());
+    Ok(())
+}
+
+/// sign for each txin
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_sign_txin(
+    address_ptr: CroAddressPtr,
+    tx_ptr: CroTxPtr,
+    which_tx_in_user: u16,
+) -> CroResult {
+    let mut user_tx: &mut CroTx = tx_ptr.as_mut().expect("get tx");
+    let address: &CroAddress = address_ptr.as_mut().expect("get address");
+    match do_cro_tx_sign_txin(&address, &mut user_tx, which_tx_in_user) {
+        Ok(_) => CroResult::success(),
+        Err(_) => CroResult::fail(),
+    }
+}
+
+fn schnorr_sign(
+    message: &TxId,
+    public_key: &PublicKey,
+    private_key: &PrivateKey,
+) -> Result<TxInWitness> {
+    let public_keys: Vec<PublicKey> = vec![public_key.clone()];
+    let multi_sig_address = MultiSigAddress::new(public_keys.to_vec(), public_keys[0].clone(), 1)?;
```

Comment:
this assumes outputs to be locked 1-of-1 -- it's ok to start with, but document this assumption

Reply:
yes, multi signature will be added